### PR TITLE
Handle paragraph alignment and direction in newline rectangles

### PR DESCRIPTION
**Pull request by alimahdiyar**
Based on #17744 and fixes flutter/flutter#53057

### CODE REVIEW

**Comment by GaryQian** [comment]
Thanks for handling the RTL case, really appreciate the extra pair of eyes on this.

This will need a test before it goes in in order to stay in sync with the currently-under-development SkParagraph by Skia. Look at `third_party/txt/tests/paragraph_unittests.cc` where you should be able to adapt on of the existing tests to verify this behavior.

cc @jason-simmons This version handles the RTL case, we should go ahead with this instead.

**Comment by alimahdiyar** [comment]
I saw the test file but I don't how to run these tests locally.
Can you tell me, please? @GaryQian

**Comment by GaryQian** [comment]
I was just mid typing when you made the changes about tracking per line. Thanks.

The tests can be run by building a host debug engine build, and running the txt_unittests executable in the engine build directory. So `./txt_unittests`

**Comment by GaryQian** [comment]
Note that the tests are meant to be run on linux, but should still be able to run on mac or windows. Some may be disabled on those platforms though due to differences in the values provided by coretext and opentype.

**Comment by alimahdiyar** [comment]
I'm getting this error
```[ RUN      ] ParagraphTest.GetRectsForRangeCenterParagraphNewlineCentered
../../flutter/third_party/txt/tests/paragraph_unittests.cc:4017: Failure
Expected equality of these values:
  boxes[0].rect.bottom()
    Which is: 73
  75
[  FAILED  ] ParagraphTest.GetRectsForRangeCenterParagraphNewlineCentered (169 ms)```
Do you know why?

**Comment by GaryQian** [comment]
What OS are you running this on? Things should pass on linux, but if it is windows or mac, we may have missed a disable here or there.

**Comment by alimahdiyar** [comment]
I'm running tests on Debian 10 Linux and it failed.
Would it be fine if we replace that line with this one in the test?
`EXPECT_FLOAT_EQ(boxes[0].rect.bottom(), paragraph->line_metrics_[1].height);`

**Comment by GaryQian** [comment]
Hmm, can you try pushing your code and seeing if the presubmit tests pass? If it does, I would just ignore it. There can be minor differences based on the exact OS you are running, which we expect. We only require maintaining a passing state across our testing infra.

File: third_party/txt/src/txt/paragraph_txt.cc [hunk 1]
```diff
@@ -1732,11 +1738,11 @@ std::vector<Paragraph::TextBox> ParagraphTxt::GetRectsForRange(
     if (line_box_metrics.find(line_number) == line_box_metrics.end()) {
       if (line.end_index != line.end_including_newline &&
           line.end_index >= start && line.end_including_newline <= end) {
-        SkScalar x = line_widths_[line_number];
-        // Move empty box to center if center aligned and is an empty line.
-        if (x == 0 && !isinf(width_) &&
-            paragraph_style_.effective_align() == TextAlign::center) {
-          x = width_ / 2;
+        SkScalar x;
+        if (glyph_lines_[line_number].positions.empty()) {
```
Comment (jason-simmons) [inline third_party/txt/src/txt/paragraph_txt.cc:1742]: Use `newline_box_positions.find(line_number)` here

Comment (alimahdiyar) [reply third_party/txt/src/txt/paragraph_txt.cc:1742]: If newline is in the selected range, it's position is evaluated in one related bidi run in the previous loop.
So we are sure about the value.
I also don't know how to handle `find()` here or what to do if it's value becomes `end()`

Comment (jason-simmons) [reply third_party/txt/src/txt/paragraph_txt.cc:1742]: Try this:
```
        auto it = newline_x_positions.find(line_number);
        if (it != newline_x_positions.end()) {
          x = it->second;
        } else {
          x = GetLineXOffset(0, line_number, false);
        }
```

If there are glyphs on the line, then `newline_x_positions` will have a value for the line and that value should be used.  If not, then use the offset for an empty line.

Comment (alimahdiyar) [reply third_party/txt/src/txt/paragraph_txt.cc:1742]: You already handled that case in the if statement.
```
        if (***glyph_lines_[line_number].positions.empty()***) {
          x = GetLineXOffset(0, line_number, false);
        } else {
          x = newline_x_positions[line_number];
        }
```

Comment (jason-simmons) [reply third_party/txt/src/txt/paragraph_txt.cc:1742]: The goal is to ensure that the code does not accidentally use a `newline_x_positions` value that was never set.

If something goes wrong and  `glyph_lines_[line_number].positions` is nonempty but `newline_x_positions[line_number]` was never set, then `newline_x_positions[line_number]` is going to yield a default value.  That will produce an unintended result.

Checking `newline_x_positions.find(line_number)` will ensure that this does not happen.

Comment (alimahdiyar) [reply third_party/txt/src/txt/paragraph_txt.cc:1742]: Done.

Comment (jason-simmons) [reply third_party/txt/src/txt/paragraph_txt.cc:1742]: The `newline_x_positions.find(line_number)` check should replace the `glyph_lines_[line_number].positions.empty()` check.  Both checks should always have the same value if everything is working as expected, but using `newline_x_positions` is safer as discussed above.

Remove the `if glyph_lines_[line_number].positions.empty()` block, and just keep the block inside the `else` below.

Comment (alimahdiyar) [reply third_party/txt/src/txt/paragraph_txt.cc]: I got it now.
Done.
Thanks for your review

**Comment by alimahdiyar** [comment]
Done

**Pull request merged**